### ISSUE TITLE
Autoupdate consuming workflows when a new composite workflow definition version is published

### ISSUE THREAD

**Issue by sfmskywalker** [elsa-workflows/elsa-core#3893]
Currently, when the user publishes a new version of a composite activity (aka sub workflow), any consuming workflows will reference the previous version, as it should be. It is then up to the user to update each consuming workflow to use the latest published version if they want to.

However, this is potentially very time consuming; the more consuming workflows, the more manual work to update each one.
To minimize the manual work, there should be an option to automatically update consuming workflows to point to the latest published version of the composite activity.

**Iteration 1**

The simplest approach is to add a checkbox to the Settings tab, called "Auto-update consuming workflows". When checked, any & all consuming workflows will be updated to reference the new version.
When this happens, there should be a notification about the number of workflow definitions affected.

**Iteration 2**

As an improvement, there could be another tab called e.g. "Consuming workflows".
This tab would display a checkbox list of consuming workflows, and a button called e.g. "Update usages".
Clicking the button will cause all selected workflows to be updated to point to the latest published version of the composite activity.

**Comment by tomy2105** [comment]
@sfmskywalker one bit of brainstorming about what can be seen in some other products/tools.

It seems (I haven't checked myself) from what you are saying is that when composite activity is included in the workflow it references it by specific version (the current one at the time of the adding I suppose).
So my question is could it be possible to do following (as addition to actions described above or as an alternative):
- have ability to choose (in the calling workflow) composite activity version being invoked (don't know if this is already possible, sorry didn't have time to check Elsa 3 yet...)
- have ability to specify "latest" as this "target" version (which would handle a certain percentage of use cases where you immediately want to use latest and not control which version is used, use cases which would be covered by your iteration 1 with the checkbox activated)

I ask this because I have a bit of dilemma about the solution you presented above.
My dilemma is what happens if the calling workflow has unpublished changes on it when you want to update the version of the composite activity it uses?
The Elsa 2 so far had linear list of versions (not a tree).
And changing the version of composite activity invoked sound like something that should mandate a new published version of the consuming workflow (same as would happen if developer changed it manually and published).
But then what would happen to the unpublished changes it has? 
It wouldn't be right to change the latest unpublished version and publish it.
It wouldn't be right to revert to last published version, update composite activity version and publish.
What you'd need probably is to save unpublished version somehow, update last published one and publish again, then update unpublished one and leave it as last unpublished version, and this sounds like a lot of work that could go wrong :(, which is why I tried to brainstorm an simpler solution that could cover some of the use cases.

Anyone, my 2 cents, I haven't been able to work with Elsa as much as I would like last couple of months so trying to help as much as I can :). Keep up the good work.

**Comment by sfmskywalker** [comment]
Thanks for the great input @tomy2105 !

You raise a valid concern with the publishing thing, and I like the suggestion to configure the activity reference itself to always point to the latest available version.

@gurkanguran What do you think?

**Comment by sfmskywalker** [comment]
@gurkanguran 

> If I understood correctly, the solution you recommend would solve the issue only if the user always(not sometimes) wants to use the latest available version of composite activity in the consuming workflow.

I think you're right. How about we offer both options? 

a) The user can update consuming workflows as per the original description.
b) The user can also configure the activity on the consuming workflow (on the Versions tab) to always use the latest published version, instead of targeting a specific one at the time the activity was added to the workfloe.

**Comment by sfmskywalker** [comment]
Great 😄

Good question. I would say, let’s not exclude those workflows, because they may contain multiple instances of the activity with different settings. Of course you could check to see if all of them want to use the latest published version always, and if that’s the case, exclude them from the list. But that might also be confusing. Unless we still show it but in a greyed out style with a tooltip explaining why it’s greyed out. But I’m not sure it’s worth it, so I’d suggest we keep it simple and reiterate as we receive feedback from real world usage.

**Comment by tomy2105** [comment]
> If I understood correctly, the solution you recommend would solve the issue only if the user always(not sometimes) wants to use the latest available version of composite activity in the consuming workflow.

Yes that was my intent. It doesn't solve the problem when he wants to explicitly control the version used. He still needs to update manually or have UI help him do it in some way.

> When it comes to the publishing dilemma, I think we should not update the composite activity version in the latest unpublished versions of consuming workflows. We should only change the reference of the composite activity it uses in the **published** workflow version without unpublishing and republishing it.

I think this is not good for two reasons.
1. Up to now all the changes to behavior of the workflow came after a publish which increased version number. Now suddenly you have a case that workflow behavior can change without that. I know it might be "nitpicking" reason but....
2. You cannot change only the **published** version. You need to change the **unpublished** one as well (if it exists), otherwise you will surely have cases where user publishes the unplublished version and unwillingly (because he forgot to update custom activity version) goes back to old version.

**Comment by tomy2105** [comment]
> The reason I said that we should not change the unpublished consuming workflow versions is that those unpublished versions should always point to the latest version of composite activities considering users would want to use the most recent version of composite activities when they are building new workflow.

That is an assumption that does not hold in many use cases :(.
What if new version of composite activity introduces some new behavior that consuming workflow is not prepared to deal with yet and will break on execution?

> However, I had a conversation with @sfmskywalker about your suggestion and we came up with a more configurable way where the user will be able to choose which composite activity version he wants to use (The latest published version or specific version). So that should fix the unpublished dilemma you mentioned.

I don't believe it does. Or I might have misunderstood you.
Imagine following scenario.
1. There is a composite activity with version 11
2. There is a consuming workflow with published version 1 consuming composite activity version 11 (not latest but specifically version 11)
3. User make a unrelated modification to consuming workflow leaving invocation of consuming activity to invoke version 11 .
4. New version of composite activity 12 is published and system automatically update published version 1 of consuming workflow (but doesn't modify unpublished version).
5. Consuming workflow now runs with version 12 of composite activity.
6. User finishes his unrelated modifications of consuming workflow and publishes (remember unpublished version is still invoking version 11 of composite activity)
7. Consuming workflow now runs, again :(, with version 11 of composite activity.

I think this scenario is bad.

**Comment by tomy2105** [comment]
> However, I think it would be better to notify the user when he wants to publish the consuming workflow that the workflow has a composite activity that references the old version in case he forgets that he chose fixed version 11. So he can update it if he wants before publishing.
> 
> Would that be a good approach you think?

I think it would work provided that we define "old version" right..... :).
The warning should ideally come if composite activity version (in my example 11) in the unpublished (soon to be published) consuming workflow version is less than the composite activity version in latest published consuming workflow version (in my example 12) which was "automatically" updated when new version of composite activity was published.

**Comment by sfmskywalker** [comment]
Yes, can you open a separate issue for that and start with a clear description of the intended UX, captured from this thread?

**Comment by sfmskywalker** [comment]
I closed this issue because the feature as originally envisioned has been implemented. We'll work on enhancing this feature using a separate issue.
